tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc

```diff
@@ -834,38 +836,66 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
 
     // Pad the non-batch dimensions of the operands to multiples of 16 as
     // required by cuBLASLt.
-    auto pad_operand = [&instr, &batch_dims](HloInstruction *&x) -> void {
-      PaddingConfig padding_config;
-      Shape padded_shape = x->shape();
-      for (int i = 0; i < x->shape().rank(); ++i) {
+    auto pad_shape = [&batch_dims](PaddingConfig &padding_config,
+                                   const Shape old_shape) {
+      Shape padded_shape = old_shape;
+      for (int i = 0; i < old_shape.rank(); ++i) {
         auto dimension = padding_config.add_dimensions();
         if (!absl::c_linear_search(batch_dims, i)) {
           int64_t padded_dimension =
-              RoundUpTo<int64_t>(x->shape().dimensions(i), 16);
+              RoundUpTo<int64_t>(old_shape.dimensions(i), 16);
           dimension->set_edge_padding_low(0);
           dimension->set_edge_padding_high(padded_dimension -
-                                           x->shape().dimensions(i));
+                                           old_shape.dimensions(i));
           dimension->set_interior_padding(0);
           padded_shape.set_dimensions(i, padded_dimension);
         }
       }
+      return padded_shape;
+    };
+
+    auto pad_operand = [&instr, &pad_shape](
+                           HloInstruction *&x,
+                           bool *is_padded = nullptr) -> void {                           
+      PaddingConfig padding_config;  
+      Shape padded_shape = pad_shape(padding_config, x->shape());
       if (!ShapeUtil::Equal(padded_shape, x->shape())) {
         HloInstruction *zero =
             instr->AddInstruction(HloInstruction::CreateConstant(
                 LiteralUtil::Zero(x->shape().element_type())));
         x = instr->AddInstruction(
             HloInstruction::CreatePad(padded_shape, x, zero, padding_config));
+        if (is_padded) {
+          *is_padded = true;
+        }
       }
       return;
     };
+
     pad_operand(a);
     pad_operand(b);
-    pad_operand(c);
+    bool c_padded = false;
+    absl::Span<const int64_t> c_dim;
+#if CUDA_VERSION > 12000
+    if (c == nullptr) {
+      c_dim = instr->shape().dimensions();
+      PaddingConfig padding_config;
+      c_padded = !ShapeUtil::Equal(pad_shape(padding_config, instr->shape()),
+                                  instr->shape());
+      c = instr->AddInstruction(
+          HloInstruction::CreateConstant(LiteralUtil::Zero(c_type)));
```

Comment:
Why create a `c` at all, if it's just going to be replaced by an empty buffer in the thunk? We don't have to pass `c` to the custom call.

Reply:
Yes, but c has to be something, in this case I made it a scalar, to get passed into HloInstruction::CreateCustomCall().

Reply:
We could create the Custom Call without c when beta is zero and CUDA is version 12 or higher.

Reply:
Yes, it's possible, but that requires lots of plumbing work and hence generating quite several compile conditional flags. A scalar placeholder operand is used here for convenience.

Reply:
I would much prefer creating the custom call without C when beta is zero. Otherwise it's inconsistent with the non-FP8 case and it's confusing how the custom call has an unused placeholder for C that is unused.

Having a lot of `CUDA_VERSION >= 12000` checks would be annoying. Is it possible to set beta to 1 for CUDA 11 when creating a zero bias, in which case we wouldn't need such checks and instead could assume we have a matrix bias iff beta is nonzero. Alternatively, perhaps we should only support this pass for CUDA 12 and above.

Reply:
If to implementing creating custom call without C to make it consistent with the non-FP8 , the operands[2] can be matrix or vector bias, check [here](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/translate/mhlo_to_lhlo_with_xla/mhlo_to_lhlo_with_xla.cc;l=971), which is what non-FP8 case does. But this plan was rejected in that PR. Check earlier discussion [here](https://github.com/tensorflow/tensorflow/pull/60020#discussion_r1145622210).